openstack/data_source_openstack_blockstorage_volume_v2.go

```diff
@@ -0,0 +1,119 @@
+package openstack
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/gophercloud/gophercloud/openstack/blockstorage/v2/volumes"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+)
+
+func dataSourceBlockStorageVolumeV2() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceBlockStorageVolumeV2Read,
+
+		Schema: map[string]*schema.Schema{
+			"region": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+
+			"name": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+
+			"status": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+
+			"metadata": {
+				Type:     schema.TypeMap,
+				Optional: true,
+				Computed: true,
+			},
+
+			// Computed values
+			"bootable": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+
+			"volume_type": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+
+			"size": {
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+
+			"source_volume_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func dataSourceBlockStorageVolumeV2Read(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	client, err := config.blockStorageV2Client(GetRegion(d, config))
+	if err != nil {
+		return fmt.Errorf("Error creating OpenStack block storage client: %s", err)
+	}
+	m := make(map[string]string)
+	for key, val := range d.Get("metadata").(map[string]interface{}) {
+		m[key] = val.(string)
+	}
+
+	listOpts := volumes.ListOpts{
+		Metadata: m,
+		Name:     d.Get("name").(string),
+		Status:   d.Get("status").(string),
+	}
+
+	allPages, err := volumes.List(client, listOpts).AllPages()
+	if err != nil {
+		return fmt.Errorf("Unable to query openstack_blockstorage_volume_v2: %s", err)
+	}
+
+	allVolumes, err := volumes.ExtractVolumes(allPages)
+	if err != nil {
+		return fmt.Errorf("Unable to retrieve openstack_blockstorage_volume_v2: %s", err)
+	}
+
+	if len(allVolumes) > 1 {
```

Comment:
do we want to support fetching all matches and failing when we expect a single match?

Reply:
I think we should only return a single result. If it is an array you can match someone else volume and I am not sure how you will be able to use an array of volumes. Also in the schema I am not sure you can return an array of objects but we can have some of the return parameters be an array.